### PR TITLE
Create unstructured Extension which can store any Extensions

### PR DESCRIPTION
**Pull request by JohnNiang**
#### What type of PR is this?

/kind feature
/kind api-change
/area core

#### What this PR does / why we need it:

This PR mainly created unstructured Extension to store any Extensions from outside and make some changes on itself.

Secondly, I refactored the Extension interface completely. Create ExtensionOperator and MetadataOperator to maintain the Extension and its metadata. So that, unstructured Extension could be implemented easily.

As a consequence, we can convert JSON/YAML string into ObjectNode and create an unstructured Extension to be created or to be updated.

There are some break changes to be noted by developers:

- Remove `Extension#metadata(Metadata metadata)`, please use `ExtensionOperator#setMetadata(MetadataOperator metadata)` instead.
- Remove `Extension#metadata()`, please use `ExtensionOperator#getMetadata()` instead.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

/cc @halo-dev/sig-halo 
/milestone 2.0

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Review by ruibaby — COMMENTED** [review]
/lgtm

**Comment by JohnNiang** [comment]
/hold
I'm going to rebase upstream/next branch to adapt the break change.

**Review by JohnNiang — COMMENTED** [review]
/unhold
/cc @halo-dev/sig-halo

File: src/test/java/run/halo/app/extension/UnstructuredTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import java.time.Instant;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+class UnstructuredTest {
+
+    ObjectMapper objectMapper = Unstructured.OBJECT_MAPPER;
+
+    String extensionJson = """
```
Comment (guqing) [inline src/test/java/run/halo/app/extension/UnstructuredTest.java:17]:
```yaml
apiVersion: v1
kind: "Role"
metadata:
  name: role-template-manage-categories
  labels:
    halo.run/role-template: true
  annotations:
    halo.run/dependencies: ["role-template-view-categories"]
    halo.run/module: "Categories Management"
    halo.run/alias-name: "Categories Management"
rules:
  - apiGroups: ["halo.run"]
    resources: ["categories"]
    verbs: ["*"]
```
形如这样的结构我是否能获取到 rules 呢，测试用例没有体现出来

Comment (JohnNiang) [reply src/test/java/run/halo/app/extension/UnstructuredTest.java:17]: > // TODO Add other convenient methods here to set and get nested fields in the future.

在代码中有一个 TODO 描述这个事情，后续会提供一个获取任意层次的 fields，不过现在没有这样的需求所以就没有提供。例如：

```java
String name = UnstructuredExtension.nestedStringField(unstructuredExtension, "metadata.name");
// or
UnstructuredExtension.setNestedStringField(unstructuredExtension, "metadata.name", "hello-halo");
```

目前 UnstructuredExtension 主要用于从 YAML 或者 JSON 加载 Extension，并将其保存或更新至数据库中，如果想要获取 rules，请用 `ExtensionClient#get(Class<E> type, String)` 得到一个确定的 Role 再进行该操作。

Comment (guqing) [reply src/test/java/run/halo/app/extension/UnstructuredTest.java:17]: > > // TODO Add other convenient methods here to set and get nested fields in the future.
> 
> 在代码中有一个 TODO 描述这个事情，后续会提供一个获取任意层次的 fields，不过现在没有这样的需求所以就没有提供。例如：
> 
> ```java
> String name = UnstructuredExtension.nestedStringField(unstructuredExtension, "metadata.name");
> // or
> UnstructuredExtension.setNestedStringField(unstructuredExtension, "metadata.name", "hello-halo");
> ```
> 
> 目前 UnstructuredExtension 主要用于从 YAML 或者 JSON 加载 Extension，并将其保存或更新至数据库中，如果想要获取 rules，请用 `ExtensionClient#get(Class<E> type, String)` 得到一个确定的 Role 再进行该操作。

ok 明白了

**Review by guqing — COMMENTED** [review]
/approve

**Pull request merged**